utils/logger.go

```diff
@@ -0,0 +1,109 @@
+// this is a new logger interface for mattermost
+
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"runtime"
+
+	l4g "github.com/alecthomas/log4go"
+)
+
+// this pattern allows us to "mock" the underlying l4g code when unit testing
+var debug = l4g.Debug
+var info = l4g.Info
+var err = l4g.Error
+
+// Logger - an instance of the log framework that can be used to log messages
+type Logger string
+
+// NewLogger - creates a new instance of Logger named after the file that created it
+func NewLogger() Logger {
```

Comment:
All good questions. The short answer is that I thought that creating a Logger instance improved readability. I had static functions before this latest commit, but because the file is in the utils class, when used, the functions lacked context about what they were doing. They just looked like `ctx = WithUserID(ctx, "some user id")` which doesn't really explain what you're adding the user id to or why.